File: apps/windows/TeslaSync.App/shared-surfaces/VehiclePaintPicker.Model.cs
using System.Globalization;
using TeslaSync.App.Core.Vehicles;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// One render-ready paint swatch — the native projection of a single mapped <c>PAINT_PALETTE_LIST</c> entry
/// (web/src/components/vehicles/VehiclePaintPicker.tsx L47-L93). Carries the palette <see cref="Id"/> (the value
/// handed back when the swatch is picked), the opaque <see cref="Swatch"/> hex the dot is filled with (web
/// <c>style={{ background: p.swatch }}</c>), the localized <see cref="Label"/> (the swatch's Narrator name, the
/// web <c>aria-label={label}</c>), the <see cref="Title"/> tooltip — the label plus the "· Auto-detected" suffix
/// for the inferred palette (web <c>title={isInferred ? `${label} · …` : label}</c>) — the <see cref="Selected"/>
/// flag (web <c>aria-checked={selected}</c>, which shows the check mark) and the <see cref="IsInferred"/> flag
/// (web <c>p.id === inferred.id</c>). Immutable so the view is a thin renderer.
/// </summary>
/// <param name="Id">The palette id (web <c>p.id</c>); the value emitted when the swatch is picked.</param>
/// <param name="Swatch">The opaque swatch hex the dot is filled with (web <c>p.swatch</c>).</param>
/// <param name="Label">The localized swatch label and Narrator name (web <c>t(p.labelKey, p.defaultLabel)</c>).</param>
/// <param name="Title">The localized tooltip (web <c>title</c>): the label, or "label · Auto-detected" when inferred.</param>
/// <param name="Selected">Whether this swatch is the active paint (web <c>selected</c> → checked + check mark).</param>
/// <param name="IsInferred">Whether this is the auto-detected palette (web <c>p.id === inferred.id</c>).</param>
public sealed record PaintSwatchItem(
    PaintPaletteId Id,
    string Swatch,
    string Label,
    string Title,
    bool Selected,
    bool IsInferred);

/// <summary>
/// Canonical metadata + localized strings + persistence identity for the vehicle-paint-picker surface — the
/// native analogue of the web <c>VehiclePaintPicker</c> (web/src/components/vehicles/VehiclePaintPicker.tsx) and
/// the storage helpers in its <c>useVehiclePaint</c> hook (web/src/hooks/useVehiclePaint.ts). It carries the
/// diagnostics slug, every render-contract i18n key/fallback the web source passes to <c>t()</c> (verbatim copy,
/// each under the catalog's <c>translation.</c> namespace so it resolves against
/// <c>apps/shared/i18n/catalog/en.json</c> + <c>apps/windows/Strings/*/Resources.resw</c> and falls back to the
/// English value headlessly), the per-palette label key map (web <c>p.labelKey</c>), the stable persistence
/// token map (web <c>PaintPaletteId</c> string ids such as <c>"red-multicoat"</c>) and the per-vehicle storage
/// key (web <c>`teslasync:vehicle:${id}:paint`</c>). UI-free so it is asserted without a XAML host.
/// </summary>
public static class VehiclePaintPickerRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "VehiclePaintPicker";

    /// <summary>i18n key for the radio-group accessible name (web <c>paint.pickerLabel</c>).</summary>
    public const string PickerLabelKey = "translation.paint.pickerLabel";

    /// <summary>English fallback for <see cref="PickerLabelKey"/> (web second arg, verbatim).</summary>
    public const string PickerLabelFallback = "Vehicle paint color";

    /// <summary>i18n key for the leading "Paint" caption (web <c>paint.label</c>).</summary>
    public const string CaptionKey = "translation.paint.label";

    /// <summary>English fallback for <see cref="CaptionKey"/> (web second arg, verbatim).</summary>
    public const string CaptionFallback = "Paint";

    /// <summary>i18n key for the auto-detected tooltip suffix (web <c>paint.detected</c>).</summary>
    public const string DetectedKey = "translation.paint.detected";

    /// <summary>English fallback for <see cref="DetectedKey"/> (web second arg, verbatim).</summary>
    public const string DetectedFallback = "Auto-detected";

    /// <summary>i18n key for the reset-to-auto button (web <c>paint.reset</c>).</summary>
    public const string ResetKey = "translation.paint.reset";

    /// <summary>English fallback for <see cref="ResetKey"/> (web second arg, verbatim).</summary>
    public const string ResetFallback = "Reset to auto-detected";

    /// <summary>i18n key for the Pearl White Multi-Coat swatch label (web <c>paint.pearlWhite</c>).</summary>
    public const string PearlWhiteKey = "translation.paint.pearlWhite";

    /// <summary>i18n key for the Midnight Silver Metallic swatch label (web <c>paint.midnightSilver</c>).</summary>
    public const string MidnightSilverKey = "translation.paint.midnightSilver";

    /// <summary>i18n key for the Deep Blue Metallic swatch label (web <c>paint.deepBlue</c>).</summary>
    public const string DeepBlueKey = "translation.paint.deepBlue";

    /// <summary>i18n key for the Solid Black swatch label (web <c>paint.solidBlack</c>).</summary>
    public const string SolidBlackKey = "translation.paint.solidBlack";

    /// <summary>i18n key for the Red Multi-Coat swatch label (web <c>paint.redMulticoat</c>).</summary>
    public const string RedMulticoatKey = "translation.paint.redMulticoat";

    /// <summary>The per-vehicle storage key prefix (web <c>STORAGE_PREFIX = 'teslasync:vehicle:'</c>).</summary>
    public const string StorageKeyPrefix = "teslasync:vehicle:";

    /// <summary>The per-vehicle storage key suffix (web <c>STORAGE_SUFFIX = ':paint'</c>).</summary>
    public const string StorageKeySuffix = ":paint";

    /// <summary>
    /// The i18n key for a palette's swatch label — the native port of the web <c>p.labelKey</c>
    /// (web/src/lib/vehicleColors.ts), under the catalog's <c>translation.</c> namespace.
    /// </summary>
    /// <param name="id">The palette id.</param>
    public static string LabelKey(PaintPaletteId id) => id switch
    {
        PaintPaletteId.MidnightSilver => MidnightSilverKey,
        PaintPaletteId.DeepBlue => DeepBlueKey,
        PaintPaletteId.SolidBlack => SolidBlackKey,
        PaintPaletteId.RedMulticoat => RedMulticoatKey,
        _ => PearlWhiteKey,
    };

    /// <summary>
    /// The stable persistence token for a palette — the native port of the web <c>PaintPaletteId</c> string id
    /// (kebab-case, e.g. <c>"red-multicoat"</c>) the hook writes to <c>localStorage</c>. Cross-compatible with
    /// the web slot so a value written by either client round-trips.
    /// </summary>
    /// <param name="id">The palette id.</param>
    public static string Token(PaintPaletteId id) => id switch
    {
        PaintPaletteId.MidnightSilver => "midnight-silver",
        PaintPaletteId.DeepBlue => "deep-blue",
        PaintPaletteId.SolidBlack => "solid-black",
        PaintPaletteId.RedMulticoat => "red-multicoat",
        _ => "pearl-white",
    };

    /// <summary>
    /// Parse a persisted token back into a palette id — the native port of the web <c>isPaintPaletteId</c>
    /// type-guard (web/src/lib/vehicleColors.ts L274): a known kebab token yields its id, anything else (a stale
    /// or foreign value) yields <see langword="null"/>, so the caller falls back to the inferred paint.
    /// </summary>
    /// <param name="token">The raw persisted token, or <see langword="null"/>.</param>
    /// <param name="id">The parsed palette id when the token is recognized.</param>
    /// <returns><see langword="true"/> when <paramref name="token"/> is a known palette token.</returns>
    public static bool TryParseToken(string? token, out PaintPaletteId id)
    {
        switch (token)
        {
            case "pearl-white":
                id = PaintPaletteId.PearlWhite;
                return true;
            case "midnight-silver":
                id = PaintPaletteId.MidnightSilver;
                return true;
            case "deep-blue":
                id = PaintPaletteId.DeepBlue;
                return true;
            case "solid-black":
                id = PaintPaletteId.SolidBlack;
                return true;
            case "red-multicoat":
                id = PaintPaletteId.RedMulticoat;
                return true;
            default:
                id = PaintPaletteId.PearlWhite;
                return false;
        }
    }

    /// <summary>
    /// The per-vehicle persistence key — the native port of the web <c>storageKey</c> helper
    /// (web/src/hooks/useVehiclePaint.ts L67): <c>teslasync:vehicle:{vehicleId}:paint</c>.
    /// </summary>
    /// <param name="vehicleId">The vehicle id (must be positive for a real slot).</param>
    public static string StorageKey(long vehicleId) =>
        string.Create(CultureInfo.InvariantCulture, $"{StorageKeyPrefix}{vehicleId}{StorageKeySuffix}");

    /// <summary>
    /// Compose the inferred swatch's tooltip — the web <c>`${label} · ${t('paint.detected', 'Auto-detected')}`</c>
    /// (web/src/components/vehicles/VehiclePaintPicker.tsx L58).
    /// </summary>
    /// <param name="label">The localized swatch label.</param>
    /// <param name="detected">The localized "Auto-detected" suffix.</param>
    public static string FormatDetectedTitle(string label, string detected)
    {
        ArgumentNullException.ThrowIfNull(label);
        ArgumentNullException.ThrowIfNull(detected);
        return string.Create(CultureInfo.CurrentCulture, $"{label} · {detected}");
    }
}

/// <summary>
/// PII-safe diagnostics for the vehicle-paint-picker surface (P1/S11 diagnostics contract). The override is a
/// browser-local cosmetic preference, not vehicle identity, and the collector records only operational signals —
/// the <see cref="RecordViewOpened"/> event with the surface slug and a <see cref="RecordPaintSelected"/> counter —
/// never the vehicle id or the chosen color. Thread-safe; mirrors the shipped surfaces' collectors.
/// </summary>
public sealed class VehiclePaintPickerDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _paintsSelected;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public VehiclePaintPickerDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of times a paint swatch (or reset) was picked.</summary>
    public long PaintsSelected => Interlocked.Read(ref _paintsSelected);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=VehiclePaintPicker</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(
            string.Create(CultureInfo.InvariantCulture, $"view.opened slug={VehiclePaintPickerRegistration.Slug}"));
    }

    /// <summary>Record that a paint was picked, emitting <c>paint.selected slug=VehiclePaintPicker</c> (no color).</summary>
    public void RecordPaintSelected()
    {
        Interlocked.Increment(ref _paintsSelected);
        _sink?.Invoke(
            string.Create(CultureInfo.InvariantCulture, $"paint.selected slug={VehiclePaintPickerRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/VehiclePaintPicker/VehiclePaintPickerModel.kt
// Pure, framework-free model + projection + diagnostics for the VehiclePaintPicker shared surface — the
// native analogue of web/src/components/vehicles/VehiclePaintPicker.tsx together with its data source
// web/src/hooks/useVehiclePaint.ts and the palette catalog web/src/lib/vehicleColors.ts. No Compose, no
// Android framework, no HTTP: every declaration here is exercised off-device in the
// :android:testReleaseUnitTest gate, keeping the composable a thin render layer.
//
// What the web source IS (and therefore the COMPLETE behaviour this surface reproduces): a small swatch row
// that lets the user override the Digital-Twin paint colour for one vehicle. It renders a `radiogroup`
// (`aria-label` = `paint.pickerLabel`) holding a short `paint.label` caption, the five fixed Tesla paint
// swatches (each a `role="radio"` button whose `aria-checked` reflects the active paint, whose `aria-label`
// is the localized paint name, and whose `title` gains a `· {paint.detected}` suffix on the auto-detected
// swatch), a live (`aria-live="polite"`) label echoing the active paint name, and — only while an override
// is in effect — a `paint.reset` text button that reverts to the auto-detected colour. The active paint is
// `override ?? inferred`, the inferred paint is derived from the Tesla `exterior_color` code, and picking the
// inferred colour explicitly CLEARS the override so the picker stays in sync if Tesla later reports a paint.
//
// What the web source is NOT: it is not a cache-then-network feed. `useVehiclePaint` resolves the active
// paint SYNCHRONOUSLY from a browser-local per-vehicle override (localStorage) layered over the inferred
// colour, with a fixed five-entry palette that is always present. There is therefore no loading / error /
// stale / offline lifecycle and no empty palette to project — modelling those would fabricate behaviour the
// web spec does not have (the same rationale the accepted VisuallyHidden / VehicleSelect local-state ports
// document). The surface's REAL states are reproduced instead and all render with no hidden region: the
// auto-detected (un-overridden) state, the overridden state (which reveals the reset affordance), the
// per-swatch selected / unselected / auto-detected-tagged states, and the "no persistable vehicle yet"
// state (id <= 0) where the picker still renders the inferred / fallback paint exactly as the web hook does.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/VehiclePaintPicker — the P3 prompt's allowed-files path) cannot form a
// valid Kotlin package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the
// package intentionally diverges from the path — exactly as the sibling VehiclePicker / VisuallyHidden
// surfaces do. `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.vehiclepaintpicker

import io.teslasync.shared.core.diagnostics.Logger

/**
 * Canonical registry metadata for the VehiclePaintPicker surface. The diagnostics [SLUG] is emitted with the
 * one-shot `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`VehiclePaintPicker`);
 * [ID] is the stable `viewModel` key the composable binds the surface with.
 */
object VehiclePaintPickerRegistration {
    /** Stable surface id (also the `viewModel` key the host binds this surface with). */
    const val ID: String = "vehicle-paint-picker"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "VehiclePaintPicker"
}

/**
 * The five stock Tesla paint ids — the native tag for the web `PaintPaletteId` union. [wireId] is the exact
 * string the web stores in `localStorage` and broadcasts across tabs (`'pearl-white'`, …); keeping it verbatim
 * means a native override is byte-compatible with the web contract and the [paintPaletteIdOf] type-guard
 * matches the web `value in PAINT_PALETTES` check 1:1.
 */
enum class PaintPaletteId(
    val wireId: String,
) {
    PearlWhite("pearl-white"),
    MidnightSilver("midnight-silver"),
    DeepBlue("deep-blue"),
    SolidBlack("solid-black"),
    RedMulticoat("red-multicoat"),
}

/**
 * One paint option the picker renders — the projection-free subset of the web `PaintPalette` the swatch row
 * actually reads (the body / surface / mirror gradients in `vehicleColors.ts` drive the VehicleTwin SVG, not
 * this picker). [labelKey] is the P1/S10 resource name; [defaultLabel] is the English fallback the web
 * `t(key, default)` call uses; [swatchArgb] is the opaque swatch colour as a 0xAARRGGBB literal (web `swatch`
 * hex), converted to a Compose `Color` only at the render boundary so this stays framework-free.
 */
data class PaintPalette(
    val id: PaintPaletteId,
    val labelKey: String,
    val defaultLabel: String,
    val swatchArgb: Long,
)

/**
 * The five stock Tesla paints in display order — the native mirror of `PAINT_PALETTE_LIST`. Adding an entry
 * here makes it appear in the picker automatically, exactly as on the web. Swatch colours are the web
 * `swatch` hexes verbatim (pearl `#e9ecf2`, midnight `#5b6675`, deep-blue `#1f3a72`, black `#0d1117`, red
 * `#a3001a`).
 */
val PAINT_PALETTE_LIST: List<PaintPalette> =
    listOf(
        PaintPalette(
            id = PaintPaletteId.PearlWhite,
            labelKey = VehiclePaintPickerKeys.PEARL_WHITE,
            defaultLabel = VehiclePaintPickerDefaults.PEARL_WHITE,
            swatchArgb = 0xFFE9ECF2,
        ),
        PaintPalette(
            id = PaintPaletteId.MidnightSilver,
            labelKey = VehiclePaintPickerKeys.MIDNIGHT_SILVER,
            defaultLabel = VehiclePaintPickerDefaults.MIDNIGHT_SILVER,
            swatchArgb = 0xFF5B6675,
        ),
        PaintPalette(
            id = PaintPaletteId.DeepBlue,
            labelKey = VehiclePaintPickerKeys.DEEP_BLUE,
            defaultLabel = VehiclePaintPickerDefaults.DEEP_BLUE,
            swatchArgb = 0xFF1F3A72,
        ),
        PaintPalette(
            id = PaintPaletteId.SolidBlack,
            labelKey = VehiclePaintPickerKeys.SOLID_BLACK,
            defaultLabel = VehiclePaintPickerDefaults.SOLID_BLACK,
            swatchArgb = 0xFF0D1117,
        ),
        PaintPalette(
            id = PaintPaletteId.RedMulticoat,
            labelKey = VehiclePaintPickerKeys.RED_MULTICOAT,
            defaultLabel = VehiclePaintPickerDefaults.RED_MULTICOAT,
            swatchArgb = 0xFFA3001A,
        ),
    )

/** The five paints keyed by id — the native mirror of `PAINT_PALETTES`. */
val PAINT_PALETTES: Map<PaintPaletteId, PaintPalette> = PAINT_PALETTE_LIST.associateBy { it.id }

/**
 * High-contrast default for cars with no `exterior_color` metadata — Pearl White, the web `FALLBACK_PAINT`.
 * It pops on the dark TeslaSync UI rather than blending into the panel like Midnight Silver would.
 */
val FALLBACK_PAINT: PaintPalette = PAINT_PALETTES.getValue(PaintPaletteId.PearlWhite)

/** Resolves a [PaintPaletteId] for its [id] palette — total over the fixed catalog. */
fun paintOf(id: PaintPaletteId): PaintPalette = PAINT_PALETTES.getValue(id)

/**
 * Narrows an arbitrary [value] (e.g. a stale persisted override) into a known [PaintPaletteId], or `null`
 * when it matches no paint — the native mirror of the web `isPaintPaletteId` type-guard. Matched against the
 * stored [PaintPaletteId.wireId] so it accepts exactly the strings the web persisted.
 */
fun paintPaletteIdOf(value: String?): PaintPaletteId? = PaintPaletteId.entries.firstOrNull { it.wireId == value }

/** Whether [value] is a known persisted paint id — the boolean form of [paintPaletteIdOf] (web type-guard). */
fun isPaintPaletteId(value: String?): Boolean = paintPaletteIdOf(value) != null

/**
 * A single forgiving Tesla `exterior_color` → paint inference rule: the [normalized] code matches when any of
 * [prefixes] is a prefix of it or any of [exacts] equals it. Data-driven so the matcher stays simple and each
 * paint's accepted spellings live in one readable place (the native shape of the web `inferPaintFromTesla`
 * `if`-ladder).
 */
private data class PaintInferenceRule(
    val prefixes: List<String>,
    val exacts: List<String>,
    val id: PaintPaletteId,
)

/** The Tesla-code → paint rules in priority order — the verbatim native port of the web `inferPaintFromTesla`. */
private val PAINT_INFERENCE_RULES: List<PaintInferenceRule> =
    listOf(
        PaintInferenceRule(prefixes = listOf("pearl"), exacts = listOf("white"), id = PaintPaletteId.PearlWhite),
        PaintInferenceRule(prefixes = listOf("midnightsilver"), exacts = listOf("silver"), id = PaintPaletteId.MidnightSilver),
        PaintInferenceRule(prefixes = listOf("deepblue"), exacts = listOf("blue", "darkblue"), id = PaintPaletteId.DeepBlue),
        PaintInferenceRule(prefixes = listOf("solidblack"), exacts = listOf("black", "obsidianblack"), id = PaintPaletteId.SolidBlack),
        PaintInferenceRule(prefixes = listOf("red"), exacts = listOf("multicoatred"), id = PaintPaletteId.RedMulticoat),
    )

/** Whether the [normalized] Tesla code satisfies [rule] — any prefix matches or any exact equals. */
private fun matchesInferenceRule(
    normalized: String,
    rule: PaintInferenceRule,
): Boolean = rule.prefixes.any(normalized::startsWith) || normalized in rule.exacts

/**
 * Maps a Tesla `exterior_color` code (e.g. `"PearlWhite"`, `"MidnightSilverMetallic"`, `"DeepBlueMetallic"`)
 * to a paint id — the exact native port of the web `inferPaintFromTesla`. The matching is forgiving:
 * case-insensitive, ignoring spaces / dashes / underscores, accepting both the bare name and the
 * `Metallic` / `MultiCoat` suffix variants Tesla emits inconsistently across models. An unknown or missing
 * code falls back to Pearl White (web `FALLBACK_PAINT`).
 */
fun inferPaintIdFromTesla(code: String?): PaintPaletteId {
    if (code.isNullOrEmpty()) return PaintPaletteId.PearlWhite
    val normalized = code.lowercase().replace(Regex("[\\s_-]"), "")
    return PAINT_INFERENCE_RULES.firstOrNull { matchesInferenceRule(normalized, it) }?.id ?: PaintPaletteId.PearlWhite
}

/**
 * Whether [vehicleId] can persist an override — the native mirror of the web `storageKey` guard
 * (`typeof id === 'number' && Number.isFinite(id) && id > 0`). A `null`, zero or negative id is "no vehicle
 * yet": persistence is disabled but the picker still renders the inferred / fallback paint, exactly as the
 * web hook does.
 */
fun isPersistableVehicleId(vehicleId: Long?): Boolean = vehicleId != null && vehicleId > 0

/**
 * Normalises a freshly-picked [picked] override against the [inferred] paint — the native mirror of the web
 * `setPaint` rule `const normalized = id === inferred.id ? null : id`. Picking the auto-detected colour (or
 * `null`) clears the override so the picker re-syncs if Tesla later reports a different paint; any other pick
 * is kept as an explicit override.
 */
fun normalizeOverride(
    picked: PaintPaletteId?,
    inferred: PaintPaletteId,
): PaintPaletteId? = if (picked == inferred) null else picked

/**
 * One swatch the picker renders, projected framework-free so the adapter is asserted off-device. [selected]
 * reflects the active paint (web `aria-checked`), [inferred] marks the auto-detected swatch (web `title`
 * suffix). [swatchArgb] is the dot colour; [labelKey] / [defaultLabel] resolve the localized paint name.
 */
data class PaintSwatch(
    val id: PaintPaletteId,
    val labelKey: String,
    val defaultLabel: String,
    val swatchArgb: Long,
    val selected: Boolean,
    val inferred: Boolean,
)

/**
 * The projected surface payload the view renders. [swatches] is the fixed five-paint row with each entry
 * tagged [PaintSwatch.selected] / [PaintSwatch.inferred]; [activeId] / [activeLabelKey] / [activeDefaultLabel]
 * are the live (`aria-live`) active-paint label; [inferredId] is the auto-detected paint; [isOverridden] is
 * true only when the user has manually picked a colour (web `isOverridden`), which gates the reset
 * affordance. The palette is always present, so there is no empty / loading shape to model.
 */
data class VehiclePaintPickerData(
    val swatches: List<PaintSwatch>,
    val activeId: PaintPaletteId,
    val activeLabelKey: String,
    val activeDefaultLabel: String,
    val inferredId: PaintPaletteId,
    val isOverridden: Boolean,
) {
    /** The active paint palette (override > inferred), never `null` over the fixed catalog. */
    val active: PaintPalette get() = paintOf(activeId)
}

/**
 * Projects the persisted [overrideId] + the Tesla [exteriorColor] onto the [VehiclePaintPickerData] the view
 * renders — the pure mirror of everything `useVehiclePaint` derives before the component returns JSX. The
 * inferred paint is [inferPaintIdFromTesla]; the active paint is `override ?? inferred` (web `paint`); each
 * swatch is tagged selected against the active paint and inferred against the auto-detected paint; and
 * [VehiclePaintPickerData.isOverridden] is `override != null` (web `overrideId !== null`).
 */
fun projectVehiclePaintPicker(
    overrideId: PaintPaletteId?,
    exteriorColor: String?,
): VehiclePaintPickerData {
    val inferredId = inferPaintIdFromTesla(exteriorColor)
    val activeId = overrideId ?: inferredId
    val active = paintOf(activeId)
    val swatches =
        PAINT_PALETTE_LIST.map { palette ->
            PaintSwatch(
                id = palette.id,
                labelKey = palette.labelKey,
                defaultLabel = palette.defaultLabel,
                swatchArgb = palette.swatchArgb,
                selected = palette.id == activeId,
                inferred = palette.id == inferredId,
            )
        }
    return VehiclePaintPickerData(
        swatches = swatches,
        activeId = activeId,
        activeLabelKey = active.labelKey,
        activeDefaultLabel = active.defaultLabel,
        inferredId = inferredId,
        isOverridden = overrideId != null,
    )
}

/**
 * Folds a swatch's localized [label] and its [inferred] flag into the swatch's accessible description — the
 * native mirror of the web `title = isInferred ? `${label} · ${detected}` : label`. The [detectedWord] is the
 * localized `paint.detected` string supplied by the render boundary so this stays framework-free; the
 * auto-detected swatch therefore announces "{paint} · Auto-detected" to TalkBack and shows the same in its
 * tooltip, while every other swatch announces just its paint name.
 */
fun paintSwatchAccessibilityLabel(
    label: String,
    inferred: Boolean,
    detectedWord: String,
): String = if (inferred) "$label · $detectedWord" else label

/**
 * The Android string-resource names the surface resolves through the i18n facade (P1/S10). Each maps a web
 * `t(key, default)` key (dots → underscores, prefixed `translation_`) onto a `translation_*` resource that
 * already ships in `values/`, `values-ar/` and `values-he/`; each name is asserted by value in the unit test.
 */
object VehiclePaintPickerKeys {
    /** Radiogroup accessible label — web `t('paint.pickerLabel', 'Vehicle paint color')`. */
    const val PICKER_LABEL: String = "translation_paint_pickerLabel"

    /** Leading caption — web `t('paint.label', 'Paint')`. */
    const val LABEL: String = "translation_paint_label"

    /** Auto-detected suffix — web `t('paint.detected', 'Auto-detected')`. */
    const val DETECTED: String = "translation_paint_detected"

    /** Reset affordance — web `t('paint.reset', 'Reset to auto-detected')`. */
    const val RESET: String = "translation_paint_reset"

    /** Pearl White paint name — web `paint.pearlWhite`. */
    const val PEARL_WHITE: String = "translation_paint_pearlWhite"

    /** Midnight Silver paint name — web `paint.midnightSilver`. */
    const val MIDNIGHT_SILVER: String = "translation_paint_midnightSilver"

    /** Deep Blue paint name — web `paint.deepBlue`. */
    const val DEEP_BLUE: String = "translation_paint_deepBlue"

    /** Solid Black paint name — web `paint.solidBlack`. */
    const val SOLID_BLACK: String = "translation_paint_solidBlack"

    /** Red Multi-Coat paint name — web `paint.redMulticoat`. */
    const val RED_MULTICOAT: String = "translation_paint_redMulticoat"
}

/** The English source strings the web `t(key, default)` calls fall back to (off-device contract). */
object VehiclePaintPickerDefaults {
    const val PICKER_LABEL: String = "Vehicle paint color"
    const val LABEL: String = "Paint"
    const val DETECTED: String = "Auto-detected"
    const val RESET: String = "Reset to auto-detected"
    const val PEARL_WHITE: String = "Pearl White Multi-Coat"
    const val MIDNIGHT_SILVER: String = "Midnight Silver Metallic"
    const val DEEP_BLUE: String = "Deep Blue Metallic"
    const val SOLID_BLACK: String = "Solid Black"
    const val RED_MULTICOAT: String = "Red Multi-Coat"
}

/** The stable, dot-namespaced diagnostics event emitted once when the surface opens (P1/S11). */
const val EVENT_VIEW_OPENED: String = "view.opened"

/** The diagnostics event emitted (PII-free) whenever the user changes the paint override. */
const val EVENT_PAINT_SET: String = "vehiclePaintPicker.set"

/** The diagnostics event emitted (PII-free) whenever the user resets to the auto-detected paint. */
const val EVENT_PAINT_RESET: String = "vehiclePaintPicker.reset"

/** Structured-log field key carrying the surface slug on every diagnostic. */
const val SURFACE_KEY: String = "surface"

/** Structured-log field key carrying the chosen paint id (a non-PII cosmetic enum value). */
const val PAINT_KEY: String = "paint"

/**
 * Emits the one PII-safe `view.opened` diagnostic carrying only the surface
 * [VehiclePaintPickerRegistration.SLUG] (P1/S11) — never a vehicle id or VIN, so a diagnostics line can never
 * leak which vehicle was viewed. Kept free of Compose so it is unit-tested with a recording [Logger]; the
 * ViewModel calls it once per surface open.
 */
fun recordVehiclePaintPickerOpened(logger: Logger) {
    logger.info(EVENT_VIEW_OPENED, mapOf(SURFACE_KEY to VehiclePaintPickerRegistration.SLUG))
}
